File: StaxLang.Interpreter/MacroTree.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Numerics;
using System.Text;
using System.Threading.Tasks;

namespace StaxLang {
    static class MacroTree {
        public class TreeNode {
            public string Description { get; }
            public string Code { get; }
            public Dictionary<char, TreeNode> Children { get; }
            public bool HasChildren => Children != null;

            public TreeNode(string code, string desc) {
                Code = code;
                Description = desc;
            }

            public TreeNode() {
                Children = new Dictionary<char, TreeNode>();
            }

            public void AddMacro(string types, string code, string desc) {
                // types: (a)rray, (b)lock, (f)raction, (i)nt, (r)eal
                if (types.Length == 0) throw new InvalidOperationException("not enough types");
                if (types.Length == 1) {
                    Children[types[0]] = new TreeNode(code, desc);
                }
                else {
                    var key = types.Last();
                    if (!Children.TryGetValue(key, out var node)) node = Children[key] = new TreeNode();
                    node.AddMacro(types.Substring(0, types.Length - 1), code, desc);
                }
            }
        }

        private static readonly IReadOnlyDictionary<char, TreeNode> MacroTrees;
        static MacroTree() {
            // types: (a)rray, (b)lock, (f)raction, (i)nt, (r)eal
            var macros = new(char alias, string types, string code, string desc)[] {
                ('!', "ii", "|g1=", "coprime?"),
                ('!', "a", "c%{[|!ms{{+k}M", "all partitionings of array"),
                ('#', "i", "1!*", "floatify"),
                ('#', "f", "1!*", "floatify"),
                ('#', "r", "", "floatify"),
                ('0', "a", "{Cim", "get indices of falsy elements"),
                ('1', "a", "{!Cim", "get indices of truthy elements"),
                ('1', "i", "2|E1#", "popcount; number of 1s in binary representation"),
                ('2', "i", "2|L@", "floor log base 2"),
                ('2', "f", "2|L@", "floor log base 2"),
                ('2', "r", "2|L@", "floor log base 2"),
                ('2', "a", "c|*", "self cross-product"),
                ('3', "a", @"Vac13|)\:fc^+|t", "rot13"),
                ('3', "ai", @"Vaca|(\:fc^+|t", "ceasar cipher"),
                ('~', "a", @"VaVA\{+kcr+|t", "toggle case"),
                ('~', "i", "c0>{|B%|2v}{U|M}?", "set all interior bits"),
                ('@', "a", "{f%", "count truthy elements"),
                ('@', "ii", "|>2%", "get the nth bit of an integer"),
                ('@', "ai", "n%%b|@aa@", "remove element at specified index; push remaining array and removed element separately"),
                ('$', "ai", "c{sM]*{{+K{+kk}{dz]}?", "cartesian power of array"),
                ('$', "a", "c{{Mm{{+K{+kk}z]?", "cartesian product of arrays in array"),
                (':', "ai", "/{hm", "get every nth element"),
                (':', "ii", "|2|^", "toggle bit"),
                (':', "ia", "{|2|^F", "toggle bits"),
                ('/', "ii", "ssn{0~{b%Csn/s,^~Wdd,}{ddVI}?", "how many times will b divide a?"),
                ('/', "ai", "n%NcN:cc0<{n%+}Mb(aat", "split array at index; push both"),
                ('/', "ia", "sn%NcN:cc0<{n%+}Mb(aat", "split array at index; push both"),
                ('/', "aa", "b[Is%^%~n,(~%t;%t,s", "split on first substring occurrence, and push both sides"),
                ('\\', "aa", @"|\{E=Cim", "get indices of differences between arrays"),
                ('=', "aa", @"|\{E=!Cim", "get indices of equal elements between arrays"),
                ('|', "a", "|ZM{|<mM", "column align - right pad columns to equal length"),
                ('_', "ii", "1!*/", "float divide"),
                ('_', "if", "1!*/", "float divide"),
                ('_', "fi", "1!*/", "float divide"),
                ('_', "ff", "1!*/", "float divide"),
                ('_', "a", "css|g1|M~{;/m,d", "reduce array; divide all elements by gcd"),
                ('*', "a", "O{*F", "array product"),
                ('+', "f", "c0>s0<-", "number sign"),
                ('+', "r", "c0>s0<-", "number sign"),
                ('+', "i", "c0>s0<-", "number sign"),
                ('+', "a", "{+}C", "prefix sums"),
                ('<', "a", "M{|<mM", "left-align columns"),
                ('>', "a", "M{|>mM", "right-align columns"),
                ('^', "a", "co=", "array is non-descending"),
                ('[', "aa", "~;%(,=", "starts with"),
                ('[', "ii", "|2Nv|&", "unset bit"),
                ('[', "ia", "{|2Nv|&F", "unset bits"),
                (']', "aa", "~;%),=", "starts with"),
                (']', "ii", "|2||", "set bit"),
                (']', "ia", "{|2||F", "set bits"),
                ('(', "a", "c%r{[|(msd", "all left rotations"),
                (')', "a", "c%r{[|)msd", "all right rotations"),
                ('{', "a", "'(s+')+", "parenthesize as string"),
                ('{', "i", "$'(s+')+", "parenthesize as string"),
                ('{', "r", "$'(s+')+", "parenthesize as string"),
                ('{', "f", "$'(s+')+", "parenthesize as string"),
                ('}', "a", "'[s+']+", "embed in square braces as string"),
                ('}', "i", "$'[s+']+", "embed in square braces as string"),
                ('}', "r", "$'[s+']+", "embed in square braces as string"),
                ('}', "f", "$'[s+']+", "embed in square braces as string"),
                ('-', "ii", "-|a", "absolute difference"),
                ('-', "ir", "-|a", "absolute difference"),
                ('-', "if", "-|a", "absolute difference"),
                ('-', "ri", "-|a", "absolute difference"),
                ('-', "rr", "-|a", "absolute difference"),
                ('-', "rf", "-|a", "absolute difference"),
                ('-', "fi", "-|a", "absolute difference"),
                ('-', "fr", "-|a", "absolute difference"),
                ('-', "ff", "-|a", "absolute difference"),
                ('-', "a", "2B{Es-m", "pairwise difference of array"),
                ('.', "a", "v\"\\b.\"{^}R", "title case string"),
                (',', "aa", "rsrs|\\r", "zip arrays;  drop prefix from longer array so end is aligned"),
                ('a', "a", "c|m|I", "indices of minima"),
                ('a', "ii", "s|Bs|z", "fixed-size zero-padded binary conversion"),
                ('A', "a", "c|M|I", "indices of maxima"),
                ('A', "i", "A|L@", "floor log base A"),
                ('A', "f", "A|L@", "floor log base A"),
                ('A', "r", "A|L@", "floor log base A"),
                ('b', "iii", "a~;>s,>!*", "value is in [range)"),
                ('b', "fii", "a~;>s,>!*", "value is in [range)"),
                ('b', "rii", "a~;>s,>!*", "value is in [range)"),
                ('b', "a", "2|E", "number from binary digits"),
                ('B', "ia", "~;%|E{;@m,d", "number in custom base"),
                ('B', "i", "2|E", "array of binary digit values"),
                ('B', "aa", "s{]ni@*mzs+{+ksd", "repeat elements respective number of times"),
                ('c', "iii", "a|m|M", "clamp integer to bounds"),
                ('c', "rii", "a|m|M1!*", "clamp float to bounds"),
                ('c', "ai", "svc^?", "set case of string based on truthiness of integer"),
                ('c', "a", "{[?k", "get first truthy element"),
                ('C', "a", "{{f|cc{Dms{hmgC", "extract columns from array of arrays"),
                ('C', "i", "~;H;|C,^/", "nth catalan number"),
                ('d', "i", "c|a{[%!fsd", "all divisors"),
                ('d', "a", "oc%vh~;t,Tc|+s%u*", "median"),
                ('D', "ai", "~;|w,|W", "trim element from both ends of string"),
                ('D', "aa", "~;|w,|W", "trim all characters from both ends of string"),
                ('e', "a", "|]{|[mc{{+k}M", "get all contiguous excerpts"),
                ('E', "a", "0U\\@", "get first and last element"),
                ('f', "a", "zs{+F", "flatten array"),
                ('f', "i", "|f|R", "prime factorization pairs: [factor exponent]"),
                ('F', "i", "|fu", "distinct prime factors"),
                ('F', "a", "{Cim", "get all falsy indices"),
                ('g', "a", "|R{hm", "remove adjacent duplicates"),
                ('g', "i", "c{2:/|2}M", "unset all but lowest bit"),
                ('G', "a", "|R{Hm", "get the lengths of runs of duplicate elements"),
                ('G', "i", "2|E|22|E", "unset all but the highest bit"),
                ('h', "a", "'<s+'>+", "embed string in angle brackets"),
                ('h', "ii", "s|Hs|z", "fixed-size zero-padded hexadecimal conversion"),
                ('i', "a", "{f{|cBB~]{f+,gC", "interleave array of arrays"),
                ('I', "aa", "{[Imsd", "get indexes of all"),
                ('J', "a", "c%|Qe~;J(,/", "squarify"),
                ('J', "ii", "JsJs", "square top 2 stack elements"),
                ('J', "ir", "JsJs", "square top 2 stack elements"),
                ('J', "if", "JsJs", "square top 2 stack elements"),
                ('J', "ri", "JsJs", "square top 2 stack elements"),
                ('J', "rr", "JsJs", "square top 2 stack elements"),
                ('J', "rf", "JsJs", "square top 2 stack elements"),
                ('J', "fi", "JsJs", "square top 2 stack elements"),
                ('J', "fr", "JsJs", "square top 2 stack elements"),
                ('J', "ff", "JsJs", "square top 2 stack elements"),
                ('m', "ai", "0|Mbs%/^a*s(", "repeat array to specified length"),
                ('m', "ii", "~;|%10?+,*", "increase to multiple"),
                ('m', "a", "cr+", "mirror (reverse and concatenate with self)"),
                ('M', "a", "cu{[]|&%i_3lm|MHsd", "mode - last to appear wins tie"),
                ('n', "ii", "vs~z]s{{cO{*F;s/c|a{[%!fsd{ns+msdmzs{+F}*{;_{/F+m,d", "all n-part factorizations"),
                ('N', "ii", "c2%{s|as}Mb|Ns#=", "integer is nth power?"),
                ('o', "aa", @"Vi|\{|Mm", "overlay; keep the maximum respective elements from two arrays"),
                ('O', "a", "co=", "is array sorted?" ),
                ('p', "i", "v{|p}{vgs", "last prime <n"),
                ('p', "a", "c{%mVI+|ms{[)msd", "truncate arrays to length of the shortest"),
                ('P', "i", "{|p}{gs", "next prime >=n"),
                ('P', "a", "c{%mVI+|ms{[(msd", "keep only the ends of arrays to match the length of the shortest"),
                ('r', "aaa", "aa/s*", "replace all substring occurrences"),
                ('r', "i", "|aNcN^|r", "centered range [-n ... n]"),
                ('R', "a", "r\"())([]][{}}{<>><\\//\\\"|t", "reflect; reverse string entire string and braces and slashes"),
                ('s', "a", "c|Ms|m-", "span of array; max - min"),
                ('S', "aa", "s-!", "is superset of"),
                ('t', "aa", "2|*|(|t", "map to next element of specified ring"),
                ('t', "i", "c|fu{u1-N*FuR", "Euler's totient"),
                ('T', "a", "{!Cim", "get all truthy indices"),
                ('T', "i", "c^*h", "triangular number (n*(n+1)/2)"),
                ('u', "a", "u%1=", "contains exactly 1 unique element?"),
                ('v', "a", "cor=", "array is non-ascending"),
                ('V', "a", "c%us|+*", "mean"),
                ('w', "a", "c1T:R+", "ascii-art palindromize; reflect braces and slashes"),
                ('W', "a", "c:R+", "ascii-art mirror; reflect braces and slashes"),
                ('x', "a", "\"[$()*+.?[\\\\\\]^{|}]\" \"\\$&\"R", "escape special chars for regex"),
            };

            var trees = new Dictionary<char, TreeNode>();
            foreach (var macro in macros) {
                if (!trees.TryGetValue(macro.alias, out var tree)) tree = trees[macro.alias] = new TreeNode();
                tree.AddMacro(macro.types, macro.code, macro.desc);
            }
            MacroTrees = trees;
        }
        public static TreeNode GetMacroTree(char alias) => MacroTrees[alias];

        private static readonly IReadOnlyDictionary<Type, char> TypeChars = new Dictionary<Type, char> {
            [typeof(List<object>)] = 'a',
            [typeof(Block)] = 'b',
            [typeof(Rational)] = 'f',
            [typeof(BigInteger)] = 'i',
            [typeof(double)] = 'r',
        };
        public static char GetTypeChar(object o) => TypeChars[o.GetType()];
    }
}

File: StaxLang.JS/macrotree.ts
import { isInt } from './integer';
import { StaxValue, isArray } from './types';
import { Block } from './block';
import { Rational } from './rational';

type NodeChildren = {[key: string]: MacroTreeNode};

export class MacroTreeNode {
    code: string;
    deprecation?: string;
    children: NodeChildren | undefined;

    hasChildren() { 
        return this.children != null;
    }

    constructor(code?: string | null, deprecation?: string | undefined) {
        if (typeof code === 'string') this.code = code;
        else this.children = {};
        this.deprecation = deprecation;
    }

    addMacro(types: string, code: string, deprecation?: string) {
        // types: (a)rray, (b)lock, (f)raction, (i)nt, (r)eal
        if (types.length === 0) throw new Error("not enough types");
        if (types.length === 1) this.children![types[0]] = new MacroTreeNode(code, deprecation);
        else {
            let key = types[types.length - 1];
            if (!(key in this.children!)) this.children![key] = new MacroTreeNode;
            this.children![key].addMacro(types.substr(0, types.length - 1), code, deprecation);
        }
    }
}

export const macroTrees: NodeChildren = {};

interface Macro {
    alias: string;
    types: string;
    code: string;
    deprecation?: string;
}

function setup() {
    const macros: Macro[] = [
        { alias: "!", types: "ii", code: "|g1="},
        { alias: "!", types: "a", code: "c%{[|!m s{{+k}M"},
        { alias: "#", types: "i", code: "1!*"},
        { alias: "#", types: "f", code: "1!*"},
        { alias: "#", types: "r", code: ""},
        { alias: "0", types: "a", code: "{Cim"},
        { alias: "1", types: "a", code: "{!Cim"},
        { alias: "1", types: "i", code: "2|E1#"},
        { alias: "2", types: "i", code: "2|L@"},
        { alias: "2", types: "f", code: "2|L@"},
        { alias: "2", types: "r", code: "2|L@"},
        { alias: "2", types: "a", code: "c|*"},
        { alias: "3", types: "a", code: "Vac13|)\\:fc^+|t"},
        { alias: "~", types: "a", code: "VaVA\\{+kcr+|t"},
        { alias: "~", types: "i", code: "c0>{|B%|2v}{UT}?"},
        { alias: "@", types: "a", code: "{f%"},
        { alias: "@", types: "ii", code: "|>2%"},
        { alias: "@", types: "ai", code: "n%%b|@aa@"},
        { alias: "$", types: "ai", code: "c{sM]*{{+K{+kk}{dz]}?"},
        { alias: "$", types: "a", code: "c{{Mm{{+K{+kk}z]?"},
        { alias: ":", types: "ai", code: "/{hm"},
        { alias: ":", types: "ii", code: "|2S"},
        { alias: ":", types: "ia", code: "{|2SF"},
        { alias: "/", types: "ii", code: "ssn{0~{b%Csn/s,^~Wdd,}{ddVI}?"},
        { alias: "/", types: "ai", code: "n%NcN:cc0<{n%+}Mb(aat", deprecation: "<code>:/</code> for split-at is deprecated.  Use <code>%</code> instead."},
        { alias: "/", types: "ia", code: "sn%NcN:cc0<{n%+}Mb(aat", deprecation: "<code>:/</code> for split-at is deprecated.  Use <code>s%</code> instead."},
        { alias: "/", types: "aa", code: "b[Is%^%~n,(~%t;%t,s"},
        { alias: "\\", types: "aa", code: "|\\{E=Cim"},
        { alias: "=", types: "aa", code: "|\\{E=!Cim"},
        { alias: "|", types: "a", code: "|ZM{|<mM", deprecation: "<code>:|</code> for column-align is deprecated.  Use <code>:&lt;</code> instead."},
        { alias: "_", types: "ii", code: "1!*/"},
        { alias: "_", types: "if", code: "1!*/"},
        { alias: "_", types: "fi", code: "1!*/"},
        { alias: "_", types: "ff", code: "1!*/"},
        { alias: "_", types: "a", code: "css|g1T~{;/m,d" },
        { alias: "*", types: "a", code: "O{*F"},
        { alias: "+", types: "f", code: "c0>s0<-"},
        { alias: "+", types: "r", code: "c0>s0<-"},
        { alias: "+", types: "i", code: "c0>s0<-"},
        { alias: "+", types: "a", code: "{+}C"},
        { alias: "<", types: "a", code: "|ZM{|<mM"},
        { alias: ">", types: "a", code: "|ZM{|>mM"},
        { alias: "^", types: "a", code: "co="},
        { alias: "[", types: "aa", code: "~;%(,="},
        { alias: "[", types: "ii", code: "|2NvI"},
        { alias: "[", types: "ia", code: "{|2NvIF"},
        { alias: "]", types: "aa", code: "~;%),="},
        { alias: "]", types: "ii", code: "|2M"},
        { alias: "]", types: "ia", code: "{|2MF"},
        { alias: "(", types: "a", code: "c%r{[|(msd"},
        { alias: ")", types: "a", code: "c%r{[|)msd"},
        { alias: "{", types: "a", code: "'(s+')+"},
        { alias: "{", types: "i", code: "$'(s+')+"},
        { alias: "{", types: "r", code: "$'(s+')+"},
        { alias: "{", types: "f", code: "$'(s+')+"},
        { alias: "}", types: "a", code: "'[s+']+"},
        { alias: "}", types: "i", code: "$'[s+']+"},
        { alias: "}", types: "r", code: "$'[s+']+"},
        { alias: "}", types: "f", code: "$'[s+']+"},
        { alias: "-", types: "ii", code: "-|a"},
        { alias: "-", types: "ir", code: "-|a"},
        { alias: "-", types: "if", code: "-|a"},
        { alias: "-", types: "ri", code: "-|a"},
        { alias: "-", types: "rr", code: "-|a"},
        { alias: "-", types: "rf", code: "-|a"},
        { alias: "-", types: "fi", code: "-|a"},
        { alias: "-", types: "fr", code: "-|a"},
        { alias: "-", types: "ff", code: "-|a"},
        { alias: "-", types: "a", code: "2B{Es-m"},
        { alias: ".", types: "a", code: "v\"\\b.\"{^}R"},
        { alias: ",", types: "aa", code: "rsrs|\\r"},
        { alias: "a", types: "a", code: "c|m|I"},
        { alias: "a", types: "ii", code: "s|Bs|z"},
        { alias: "A", types: "a", code: "c|M|I"},
        { alias: "A", types: "i", code: "A|L@"},
        { alias: "A", types: "f", code: "A|L@"},
        { alias: "A", types: "r", code: "A|L@"},
        { alias: "b", types: "iii", code: "a~;>s,>!*"},
        { alias: "b", types: "fii", code: "a~;>s,>!*"},
        { alias: "b", types: "rii", code: "a~;>s,>!*"},
        { alias: "b", types: "a", code: "2|E"},
        { alias: "B", types: "ia", code: "~;%|E{;@m,d"},
        { alias: "B", types: "i", code: "2|E"},
        { alias: "B", types: "aa", code: "s{]ni@*mzs+{+ksd"},
        { alias: "c", types: "iii", code: "atT"},
        { alias: "c", types: "rii", code: "atT1!*"},
        { alias: "c", types: "ai", code: "svc^?"},
        { alias: "c", types: "a", code: "{[?k"},
        { alias: "C", types: "a", code: "{{f|cc{Dms{hmgC" },
        { alias: "C", types: "i", code: "~;H;|C,^/"},
        { alias: "d", types: "i", code: "c|a{[%!fsd"},
        { alias: "d", types: "a", code: "oc%vh~;t,Tc|+s%u*"},
        { alias: "D", types: "ai", code: "~;|w,|W"},
        { alias: "D", types: "aa", code: "~;|w,|W"},
        { alias: "e", types: "a", code: "|]{|[mc{{+k}M"},
        { alias: "E", types: "a", code: "0U\\@"},
        { alias: "f", types: "a", code: "zs{+F"},
        { alias: "f", types: "i", code: "|f|R"},
        { alias: "F", types: "i", code: "|fu"},
        { alias: "F", types: "a", code: "{Cim", deprecation: "<code>:F</code> for falsy indices is deprecated.  Use <code>:0</code> instead."},
        { alias: "g", types: "a", code: "|R{hm"},
        { alias: "g", types: "i", code: "c{2:/|2}M"},
        { alias: "G", types: "a", code: "|R{Hm"},
        { alias: "G", types: "i", code: "2|E|22|E"},
        { alias: "h", types: "a", code: "'<s+'>+"},
        { alias: "h", types: "ii", code: "s|Hs|z"},
        { alias: "i", types: "a", code: "{f{|cBB~]{f+,gC"},
        { alias: "I", types: "aa", code: "{[Imsd"},
        { alias: "J", types: "a", code: "c%|Qe~;J(,/"},
        { alias: "J", types: "ii", code: "JsJs"},
        { alias: "J", types: "ir", code: "JsJs"},
        { alias: "J", types: "if", code: "JsJs"},
        { alias: "J", types: "ri", code: "JsJs"},
        { alias: "J", types: "rr", code: "JsJs"},
        { alias: "J", types: "rf", code: "JsJs"},
        { alias: "J", types: "fi", code: "JsJs"},
        { alias: "J", types: "fr", code: "JsJs"},
        { alias: "J", types: "ff", code: "JsJs"},
        { alias: "m", types: "ai", code: "0Tbs%/^a*s("},
        { alias: "m", types: "ii", code: "~;|%10?+,*"},
        { alias: "m", types: "a", code: "cr+"},
        { alias: "M", types: "a", code: "cu{[]|&%i_3lm|MHsd"},
        { alias: "n", types: "ii", code: "vs~z]s{{cO{*F;s/c|a{[%!fsd{ns+msdmzs{+F}*{;_{/F+m,d"},
        { alias: "N", types: "ii", code: "c2%{s|as}Mb|Ns#="},
        { alias: "o", types: "aa", code: "Vi|\\{|Mm"},
        { alias: "p", types: "i", code: "v{|p}{vgs"},
        { alias: "p", types: "a", code: "c{%mVI+|ms{[)msd"},
        { alias: "P", types: "i", code: "{|p}{gs"},
        { alias: "P", types: "a", code: "c{%mVI+|ms{[(msd"},
        { alias: "r", types: "aaa", code: "aa/s*"},
        { alias: "r", types: "i", code: "|aNcN^|r"},
        { alias: "R", types: "a", code: "r\"())([]][{}}{<>><\\//\\\"|t"},
        { alias: "s", types: "a", code: "c|Ms|m-"},
        { alias: "S", types: "aa", code: "s-!"},
        { alias: "t", types: "aa", code: "2|*|(|t"},
        { alias: "t", types: "i", code: "c|fu{u1-N*FuR"},
        { alias: "T", types: "a", code: "{!Cim", deprecation: "<code>:T</code> for truthy indices is deprecated.  Use <code>:1</code> instead."},
        { alias: "T", types: "i", code: "c^*h", deprecation: "<code>:T</code> for triangular numbers is deprecated.  Use <code>|+</code> instead."},
        { alias: "u", types: "a", code: "u%1="},
        { alias: "v", types: "a", code: "cor="},
        { alias: "V", types: "a", code: "c%us|+*"},
        { alias: "w", types: "a", code: "c1T:R+"},
        { alias: "W", types: "a", code: "c:R+"},
        { alias: "x", types: "a", code: "\"[$()*+.?[\\\\\\]^{|}]\" \"\\$&\"R"},
    ];

    for (let macro of macros) {
        if (!(macro.alias in macroTrees)) macroTrees[macro.alias] = new MacroTreeNode;
        macroTrees[macro.alias].addMacro(macro.types, macro.code, macro.deprecation);
    }
}
setup();

export function getTypeChar(o: StaxValue): string {
    if (isArray(o)) return 'a';
    if (o instanceof Block) return 'b';
    if (o instanceof Rational) return 'f';
    if (isInt(o)) return 'i';
    if (typeof o === 'number') return 'r';
    throw new Error('no type char for unknown type');
}